Add const generics to infer (and transitive dependencies)

varkor opened the pull request:
Split out from #53645. This work is a collaborative effort with @yodaldevoid.

There are a number of stubs. These are mainly to ensure we don't overlook them when completing the implementation, but are not necessary for the initial implementation. We plan to address these in follow up PRs.

r? @eddyb / @nikomatsakis

eddyb commented on src/librustc/ich/impls_ty.rs:
```diff
@@ -1121,6 +1121,7 @@ impl_stable_hash_for!(enum infer::canonical::CanonicalVarKind {
     PlaceholderTy(placeholder),
     Region(ui),
     PlaceholderRegion(placeholder),
+    Const(ui),
```
You likely need `PlaceholderConst` too.

varkor commented on src/librustc/infer/canonical/mod.rs:
```diff
@@ -443,8 +460,19 @@ impl<'tcx> CanonicalVarValues<'tcx> {
+                            ty::LazyConst::Unevaluated(def_id, _) => {
+                                tcx.type_of(*def_id)
+                            }
+                            ty::LazyConst::Evaluated(ty::Const { ty, .. }) => ty,
```
Rebasing is such a pain.

eddyb replied:
Sure, but this PR makes very little sense atm anywhere "lazy" shows up. We'd have to fix it either before or after, and I want to be able to review this PR instead of just skimming it.

oli-obk replied:
Ok, I'll create a PR cleaning up the situation

crlf0710 replied:
#59178 is merged.

varkor commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,271 @@
+use rustc_data_structures::snapshot_vec as sv;
+use rustc_data_structures::unify as ut;
+
+pub struct ConstVariableTable<'tcx> {
```
We had something more basic initially, but kept running into situations where an analogous abstraction to `TypeVariableTable` didn't exist. Replicating a similar data structure for consts made everything clearer, but it's possible that some of what we were trying to do is simply unnecessary at that level of generality for consts.

> The problem with types is that there are "integer infer variables" and "float infer variables", but constants only have one kind of inference variable.

Is this not a separate concern (i.e. there's `type_variables`, `int_unification_table` and `float_unification_table` and at the moment, `const_variables` follows the structure laid out by `type_variables` rather than the other two)? It might be easiest to identify places where the const unification table is doing more work than necessary, so we can work out if it isn't actually necessary.

eddyb replied:
Oh I just thought there is already something like this, then I don't know.

Is there no `..._table` field that has a type that's... not from librustc? Some general-purposes "unification table"?

varkor replied:
I don't see one in `InferCtxt` at least. It could make sense to create an abstraction if it doesn't exist (and I'm not just overlooking it somewhere) as a `C-cleanup` issue afterwards.

eddyb replied:
I just checked and it turns out that I was referring to `int_unification_table` and `float_unification_table`, which rely on e.g.:
https://github.com/rust-lang/rust/blob/aa97448caf49ee25271a5ad58482283404b9144f/src/librustc/infer/unify_key.rs#L8-L13

The reason this wasn't immediately obvious was that you have a `values` field and that it *doesn't* hold the actual variable values (maybe it used to, for `TypeVariableTable`, at some point?) *but it's really* "origins" (and should be named as such; in `TypeVariableTable` too).

Note that the `UnificationTable` holds `<Vid as UnifyKey>::Value` for each active vid, and you could combine `ConstVariableData` and `ConstVariableValue` into one type to use as that `Value` associated type.

A bit surprised to see integer/float inference variables don't have an origin attached, but they could all use the same mechanism.

IMO the reason `type_variables` exists as a separate module is mainly because of subtyping relations, and we shouldn't need that here.

eddyb commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,271 @@
+    MiscVariable(Span),
+    ConstInference(Span),
+    ConstParameterDefinition(Span, InternedString),
+    SubstitutionPlaceholder(Span),
```
I just realized all of these have `Span`s, maybe that can be factored out? (but this is probably not the only place)

varkor replied:
I'll refactor `TypeVariableOrigin` and `ConstVariableOrigin` in a follow up.

varkor replied:
Opened https://github.com/rust-lang/rust/pull/59331.

eddyb commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,271 @@
+pub enum ConstVariableValue<'tcx> {
+    Known { value: &'tcx ty::LazyConst<'tcx> },
+    Unknown { universe: ty::UniverseIndex },
+}
```
@nikomatsakis I'm thinking maybe there should be a generic `enum` like this to share between types and consts.

nikomatsakis replied:
Yeah, so, I rewrote ena for lark and I made a bunch of changes. As I recall, one of the things I changed was eliminating this sort of boilerplate (I had some other goals too, like being able to track the tree of unifications for better error messages). To be honest, I sort of *forget* what I did in Lark. I think this is another work item that is worth trying to postpone till after the PR lands, and then cleanup separately.

eddyb commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,271 @@
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableValue<'tcx> {
+    Known { value: &'tcx ty::LazyConst<'tcx> },
+    Unknown { universe: ty::UniverseIndex },
```
Is this enough? Don't we also need the type?

nikomatsakis replied:
Well, I think the type is carried in the `ty::Const` wrapper, for better or worse:

```rust
    #[inline]
    pub fn mk_const_var(self, v: ConstVid<'tcx>, ty: Ty<'tcx>) -> &'tcx Const<'tcx> {
        self.mk_const(ty::Const {
            val: ConstValue::Infer(InferConst::Var(v)),
            ty,
        })
    }
```

varkor replied:
It's not carried around in `Unknown`, which I think is the part @eddyb was referring to. (There's some other discussion about this elsewhere, but there's a `FIXME` in the code about this now.)

eddyb commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,271 @@
+    fn tag() -> &'static str { "ConstVid" }
+}
+
+impl<'tcx> ut::UnifyValue for ConstVariableValue<'tcx> {
```
Yeah, this entire `impl` is just universal boilerplate.

nikomatsakis replied:
Yep. As above, I think we ought to land as is, but we should definitely make some time to talk about how to do this better.

eddyb commented on src/librustc/infer/canonical/mod.rs:
```diff
@@ -414,8 +414,17 @@ impl<'cx, 'gcx, 'tcx> InferCtxt<'cx, 'gcx, 'tcx> {
+                self.tcx.mk_lazy_const(ty::LazyConst::Evaluated(
+                    ty::Const {
+                        val: ConstValue::Placeholder(placeholder_mapped),
+                        ty: self.tcx.types.err, // FIXME(const_generics)
```
Sounds like `PlaceholderConst` should contain the type.

varkor replied:
Yeah, it probably should. The reason I didn't do it earlier was it ended up propagating `<'tcx>` everywhere and I wanted to make sure it was definitely necessary before going down that route.

yodaldevoid replied:
I attempted propagating `<'tcx>` around when originally adding `CanonicalVarKind::Const` and found it to be impossible because `CanonicalVarInfo`s are allocated in the global arena.

eddyb replied:
Sound like we need to bother @nikomatsakis.

But also, aren't const types supposed to be always be "global"?

yodaldevoid replied:
AFAIK, types are only global after they have been lifted and their dependencies on type parameters and such are taken care of.

eddyb replied:
Type parameters are "global" in this sense, only inference variables aren't.
I'm saying that I don't think we should ever have a constant value in the type system with an *uninferred type*.

nikomatsakis replied:
Hmm. Something feels wrong here indeed. I'm trying too put my finger on what it is. I guess in short I don't like the idea of there being a **type** in the canonical-var-info -- I feel like types should **in the canonicalized value**.

I think what I would expect is that the placeholder doesn't replace the entire `ty::Const`, but rather **just** the `ConstValue`. In this way, it can be instantiated without needing the type. 

(Similarly, in the existential case that appears above, you wouldn't make a "Fresh type variable" to represent its type -- that also seems wrong.)

bors commented:
:umbrella: The latest upstream changes (presumably #56864) made this pull request unmergeable. Please resolve the merge conflicts.

varkor commented:
This is currently blocked on https://github.com/rust-lang-nursery/ena/pull/19 and response from @nikomatsakis regarding the review comments above.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/185803656) ([raw log](https://api.travis-ci.com/v3/job/185803656/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:057be65f:start=1552942958137602078,finish=1552942960617471838,duration=2479869760
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
Setting environment variables from .travis.yml
---
[00:32:25]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:32:26] error: incorrect close delimiter: `}`
[00:32:26]    --> src/librustc/ty/relate.rs:637:13
[00:32:26]     |
[00:32:26] 630 |             ConstValue::Unevaluated(b_def_id, b_substs)) if a_def_id == b_def_id => {
[00:32:26] ...
[00:32:26] ...
[00:32:26] 633 |                 Ok(tcx.mk_const(ty::Const {
[00:32:26]     |                   - un-closed delimiter
[00:32:26] 637 |             }
[00:32:26]     |             ^ incorrect close delimiter
[00:32:26] 
[00:33:09] error: aborting due to previous error
---
travis_time:end:08468a57:start=1552944961548020918,finish=1552944961552755370,duration=4734452
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:14182040
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:09296e1d
travis_time:start:09296e1d
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:1799bc1f
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

crlf0710 commented:
ena 0.12.0 is [released](https://docs.rs/ena/0.12.0/ena/).

varkor commented:
I missed a detail in the previous pull request to ena: this is now blocked on https://github.com/rust-lang-nursery/ena/pull/21.

bors commented:
:umbrella: The latest upstream changes (presumably #58305) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@varkor This should be unblocked now, just released `ena v0.13`.

varkor commented:
This is now blocked on https://github.com/rust-lang/rust/pull/59415.

varkor commented:
I've addressed all the comments that aren't waiting on feedback from @nikomatsakis.

Centril commented:
Reassigning the remainder of the review to r? @nikomatsakis then.

eddyb commented:
Anyone else from @rust-lang/wg-traits want to take a look at the questions I left for @nikomatsakis, too?

WiSaGaN commented:
Does this still block on @nikomatsakis and @rust-lang/wg-traits?

crlf0710 commented:
@WiSaGaN Yes, althrough niko has responded that he'll add it to his list last week:
https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/.2359008.20changes.20to.20infer

nikomatsakis commented:
My great apologies for the delay. I've scheduled this for review this coming Wednesday at 12:00 UTC-04:00.

varkor commented on src/librustc/ty/relate.rs:
```diff
@@ -581,6 +581,70 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
+    match (a, b) {
+        (ty::LazyConst::Evaluated(a_eval), ty::LazyConst::Evaluated(b_eval)) => {
+            // Only consts whose types are equal should be compared.
+            assert_eq!(a_eval.ty, b_eval.ty);
```
Ah, that does seem plausible. I'll remove this assertion.

varkor commented on src/librustc/infer/const_variable.rs:
```diff
@@ -0,0 +1,242 @@
+                &ConstVariableValue::Known { value: value1 },
+                &ConstVariableValue::Known { value: value2 }
+            ) => {
+                match <&'tcx ty::LazyConst<'tcx>>::unify_values(&value1, &value2) {
```
Not as far as I know. This was probably an oversight.

varkor commented on src/librustc/infer/canonical/canonicalizer.rs:
```diff
@@ -633,4 +633,34 @@ impl<'cx, 'gcx, 'tcx> Canonicalizer<'cx, 'gcx, 'tcx> {
+        if bound_to != const_var {
+            self.fold_const(bound_to)
+        } else {
+            let ty = match const_var {
```
I think this code has been refactored since.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/193837206) ([raw log](https://api.travis-ci.com/v3/job/193837206/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0e602870:start=1555538071731456356,finish=1555538073903982712,duration=2172526356
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
  Downloading https://files.pythonhosted.org/packages/db/c8/7dcf9dbcb22429512708fe3a547f8b6101c0d02137acbd892505aee57adf/colorama-0.3.9-py2.py3-none-any.whl
Collecting futures<4.0.0,>=2.2.0; python_version == "2.6" or python_version == "2.7" (from s3transfer<0.3.0,>=0.2.0->awscli)
  Downloading https://files.pythonhosted.org/packages/2d/99/b2c4e9d5a30f6471e410a146232b4118e697fa3ffc06d6a65efde84debd0/futures-3.2.0-py2-none-any.whl
Collecting urllib3<1.25,>=1.20; python_version == "2.7" (from botocore==1.12.132->awscli)
  Downloading https://files.pythonhosted.org/packages/df/1c/59cca3abf96f991f2ec3131a4ffe72ae3d9ea1f5894abe8a9c5e3c77cfee/urllib3-1.24.2-py2.py3-none-any.whl (131kB)
    15% |█████                           | 20kB 26.3MB/s eta 0:00:01
    23% |███████▌                        | 30kB 32.1MB/s eta 0:00:01
    31% |██████████                      | 40kB 34.8MB/s eta 0:00:01
    38% |████████████▍                   | 51kB 36.6MB/s eta 0:00:01
---
[00:08:09]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:08:46] error: unused variable: `value1`
[00:08:46]    --> src/librustc/infer/unify_key.rs:134:52
[00:08:46]     |
[00:08:46] 134 |                 ConstVariableValue::Known { value: value1 },
[00:08:46]     |                                                    ^^^^^^ help: consider prefixing with an underscore: `_value1`
[00:08:46]     = note: `-D unused-variables` implied by `-D warnings`
[00:08:46] 
[00:08:46] error: unused variable: `value2`
[00:08:46]    --> src/librustc/infer/unify_key.rs:135:52
[00:08:46]    --> src/librustc/infer/unify_key.rs:135:52
[00:08:46]     |
[00:08:46] 135 |                 ConstVariableValue::Known { value: value2 }
[00:08:46]     |                                                    ^^^^^^ help: consider prefixing with an underscore: `_value2`
[00:08:47] error: aborting due to 2 previous errors
[00:08:47] 
[00:08:47] error: Could not compile `rustc`.
[00:08:47] 
---
205996 ./obj/build/cache/2019-04-11
157492 ./obj/build/bootstrap/debug/incremental
156496 ./src/llvm-project/clang
142508 ./obj/build/bootstrap/debug/incremental/bootstrap-hfsog967tquc
142504 ./obj/build/bootstrap/debug/incremental/bootstrap-hfsog967tquc/s-fbe4eia0b3-5jds3l-3k7gym5px36ja
108532 ./src/llvm-project/lldb
106696 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc
97584 ./src/llvm-project/clang/test
96736 ./.git
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented:
OK, unfortunately I got too busy today and failed to do the review I had planned to do. I've rescheduled that to Tuesday (tomorrow) at 16:00 UTC-04:00.

bors commented:
:umbrella: The latest upstream changes (presumably #60211) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc/infer/combine.rs:
```diff
@@ -107,13 +109,68 @@ impl<'infcx, 'gcx, 'tcx> InferCtxt<'infcx, 'gcx, 'tcx> {
+                // All other cases of inference with other variables are errors.
+                (ConstValue::Infer(InferConst::Var(_)), ConstValue::Infer(_)) |
+                (ConstValue::Infer(_), ConstValue::Infer(InferConst::Var(_))) => {
+                    bug!("tried to combine ConstValue::Infer/ConstValue::Infer(InferConst::Var)")
```
Well, the other variants of `InferConst` are:

```rust
/// An inference variable for a const, for use in const generics.
#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd,
         Ord, RustcEncodable, RustcDecodable, Hash, HashStable)]
pub enum InferConst<'tcx> {
    /// Infer the value of the const.
    Var(ConstVid<'tcx>),
    /// A fresh const variable. See `infer::freshen` for more details.
    Fresh(u32),
    /// Canonicalized const variable, used only when preparing a trait query.
    Canonical(DebruijnIndex, BoundVar),
}
```

so, yes, this is due to the placeholder stuff. I don't know whether it will get "cleaned up" or not, I guess that depends a bit on what the final shape of our type story looks like, which is something I've been thinking a lot about (I'd like to share the code between rustc and chalk, for example). 

I think overall I'd like to move away from the "zoo" of variants that exist today and have three:

- free variables: inference variables (existentially quantified) and placeholders (universally quantified)
- bound variables (based on debruijn indices): used for canonicalization, `for<'a>` types, etc.

but that's probably a larger conversation than just this PR.

nikomatsakis replied:
I think in short this part of the code seems ok to land as is, it seems comparable to the other cases, and we can refactor them together.

nikomatsakis commented on src/librustc/infer/mod.rs:
```diff
@@ -918,17 +920,17 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
-            let a = self.shallow_resolve(predicate.skip_binder().a);
-            let b = self.shallow_resolve(predicate.skip_binder().b);
+            let a = self.shallow_resolve_type(predicate.skip_binder().a);
+            let b = self.shallow_resolve_type(predicate.skip_binder().b);
```
It seems like it could work on any type-foldable.

nikomatsakis replied:
It's not something you would normally want to do on a composite structure, but I don't see the *harm* in doing so -- presumably the only actual callers would be people giving in a `Ty` or `Const` directly.

nikomatsakis replied:
So in summary: yes, I think we should make it generic, rather than introduce multiple variants with different names.

nikomatsakis commented on src/librustc/ty/relate.rs:
```diff
@@ -575,6 +585,70 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
+                }
+            }
+        }
+        // FIXME(const_generics): this is probably wrong (regarding TyProjection)
```
Well, to be quite honest, I'm a bit confused by what the "normalization policy" is here. If this were doing eager normalization, as I would expect at the moment, then this doesn't seem wrong. But if it's lazy normalization, then indeed it is wrong. It seems like we can land it as is, but I think that overall it would be good to try and document and talk out the design of constants and const generics.

nikomatsakis replied:
Design meeting! :)

nikomatsakis commented:
OK, I think I answered all of the outstanding comments from @eddyb. I've read through enough of this PR that I feel comfortable with it landing overall -- but I think I would really like to have some time to talk out the overall strategy, especially around normalization. 

@varkor and @yodaldevoid -- maybe we can sync up a bit next week and you can try to talk me through it? (Over Zulip would be fine.) I'm interested in having the design here [written out and submitted as a proposal](https://github.com/rust-lang/compiler-team/pull/75). Anyway, I'll ping you on Zulip to discuss.

As far as the concerns @eddyb raised, I think they fall into two categories:

- Boilerplate around inference variables: I think we should table this concern but revisit it in a uniform way outside of this PR.
- The integration of const values into canonicalization: [Something does indeed seem wrong here](https://github.com/rust-lang/rust/pull/59008#discussion_r279094359), though I'm not sure if it should block this PR per s, as I guess none of this is "exposed" to end useres.

varkor commented:
I've fixed the remaining issues regarding refactoring (that are in-scope for this PR) and rebased. As far as I can tell now, there's just the question of whether we want to settle the questions about canonicalisation before we merge this PR, or whether we can leave it for afterwards?

bors commented:
:umbrella: The latest upstream changes (presumably #60195) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
Broken canonicalization will probably prevent MIR borrowck from working, but we might not need that yet.

r=me unless @nikomatsakis disagrees

nikomatsakis commented:
I think we can settle this stuff afterwards and am fine with merging

nikomatsakis commented:
Well, wait, will this potentially affect things like `[u8; 22]`? I'm assuming not

varkor commented:
It shouldn't affect arrays, no (and I'm not aware that it does).

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit a68ed060bdab2d4c5d36062cda24f89d22488271 has been approved by `eddyb`

<!-- @bors r=eddyb a68ed060bdab2d4c5d36062cda24f89d22488271 -->

bors commented:
:hourglass: Testing commit a68ed060bdab2d4c5d36062cda24f89d22488271 with merge 92b5e20ad59b183a2e6cd79423f54b8aa7ac9cbf...
<!-- homu: {"type":"BuildStarted","head_sha":"a68ed060bdab2d4c5d36062cda24f89d22488271","merge_sha":"92b5e20ad59b183a2e6cd79423f54b8aa7ac9cbf"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/110323991), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/24247418)
Approved by: eddyb
Pushing 92b5e20ad59b183a2e6cd79423f54b8aa7ac9cbf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"eddyb","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/24247418","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/110323991"},"merge_sha":"92b5e20ad59b183a2e6cd79423f54b8aa7ac9cbf"} -->
